lib/services/local/headlessauthn_watcher_test.go

```diff
@@ -26,171 +26,324 @@ import (
+		require.Nil(t, <-headlessAuthnCh)
 	})
 
-	t.Run("WaitCompareAndSwap", func(t *testing.T) {
-		stub, err := identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
+	t.Run("WaitBackend", func(t *testing.T) {
+		t.Parallel()
+		s := newSuite(t)
+
+		stub, err := s.identity.CreateHeadlessAuthenticationStub(ctx, pubUUID)
 		require.NoError(t, err)
-		t.Cleanup(func() { require.NoError(t, identity.DeleteHeadlessAuthentication(ctx, pubUUID)) })
 
-		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*2)
+		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*5)
 		defer waitCancel()
 
-		headlessAuthnCh, errC := waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			return ha.State == types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED, nil
+		// Wait should immediately check the backend and return the existing headless authentication stub.
+		headlessAuthn, err := s.watcher.Wait(waitCtx, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
+			return true, nil
 		})
 
-		replace := *stub
-		replace.State = types.HeadlessAuthenticationState_HEADLESS_AUTHENTICATION_STATE_APPROVED
-		replace.PublicKey = []byte(sshPubKey)
-		replace.User = "user"
-
-		swapped, err := identity.CompareAndSwapHeadlessAuthentication(ctx, stub, &replace)
 		require.NoError(t, err)
+		require.Equal(t, stub, headlessAuthn)
+	})
 
-		require.NoError(t, <-errC)
-		require.Equal(t, swapped, <-headlessAuthnCh)
+	t.Run("WaitTimeout", func(t *testing.T) {
+		t.Parallel()
+		s := newSuite(t)
+
+		waitCtx, waitCancel := context.WithTimeout(ctx, time.Millisecond*10)
+		defer waitCancel()
+
+		_, err := s.watcher.Wait(waitCtx, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) { return true, nil })
+		require.Error(t, err)
+		require.Equal(t, waitCtx.Err(), err)
 	})
 
 	t.Run("StaleCheck", func(t *testing.T) {
-		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*2)
-		defer waitCancel()
+		t.Parallel()
+		s := newSuite(t)
 
-		// Create two waiters - a blocked consumer and a free consumer.
-		blockWait := make(chan struct{})
-		_, errC := waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			<-blockWait
-			return false, nil
-		})
+		waitCtx, waitCancel := context.WithTimeout(ctx, time.Second*5)
+		defer waitCancel()
 
-		notifyReceived := make(chan struct{}, 1)
-		waitInGoroutine(waitCtx, t, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
-			notifyReceived <- struct{}{}
+		// Create a waiter that we can block/unblock.
+		firstEmitReceived := make(chan struct{})
+		blockWaiter := make(chan struct{})
+		_, blockedWaiterErrC := waitInGoroutine(waitCtx, t, s.watcher, pubUUID, func(ha *types.HeadlessAuthentication) (bool, error) {
+			select {
+			case <-firstEmitReceived:
+			default:
+				close(firstEmitReceived)
+				// we only block the first event received, incase additional events get to this waiter.
```

Comment:
This isn't blocking during a receive, is it?

(also typo on "incase".)

Could we simplify the entire inner function to:

```go
				close(firstEmitReceived)
				<-blockWaiter
```